integration-tests/byoid/integration_test.go

```diff
@@ -231,3 +236,104 @@ func TestURLBasedCredentials(t *testing.T) {
 		},
 	})
 }
+
+// Tests to make sure AWS based external credentials work properly.
+func TestAWSBasedCredentials(t *testing.T) {
+	data := url.Values{}
+	data.Set("audience", clientID)
+	data.Set("includeEmail", "true")
+
+	client, err := google.DefaultClient(context.Background(), "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		t.Fatalf("Failed to create default client: %v", err)
+	}
+	resp, err := client.PostForm(fmt.Sprintf("https://iamcredentials.googleapis.com/v1/%s:generateIdToken", clientID), data)
+	if err != nil {
+		t.Fatalf("Failed to generate an ID token: %v", err)
+	}
+	if resp.StatusCode != 200 {
+		t.Fatalf("Failed to get Google ID token for AWS test: %v", err)
+	}
+
+	var res map[string]interface{}
+
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		t.Fatalf("Could not successfully parse response from generateIDToken: %v", err)
+	}
+	token, ok := res["token"]
+	if !ok {
+		t.Fatalf("Didn't receieve an ID token back from generateIDToken")
+	}
+
+	data = url.Values{}
+	data.Set("Action", "AssumeRoleWithWebIdentity")
+	data.Set("Version", "2011-06-15")
+	data.Set("DurationSeconds", "3600")
+	data.Set("RoleSessionName", os.Getenv("GCLOUD_TESTS_GOLANG_AWS_ROLE_NAME"))
+	data.Set("RoleArn", os.Getenv("GCLOUD_TESTS_GOLANG_AWS_ROLE_ID"))
+	data.Set("WebIdentityToken", token.(string))
+
+	resp, err = http.PostForm("https://sts.amazonaws.com/", data)
+	if err != nil {
+		t.Fatalf("Failed to post data to AWS: %v", err)
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatalf("Failed to parse response body from AWS: %v", err)
+	}
+
+	type Credentials struct {
+		SessionToken    string `xml:"SessionToken"`
+		SecretAccessKey string `xml:"SecretAccessKey"`
+		AccessKeyID     string `xml:"AccessKeyId"`
+	}
+	type AssumeRoleWithWebIdentityResult struct {
+		Credentials Credentials `xml:"Credentials"`
+	}
+	type AssumeRoleWithWebIdentityResponse struct {
+		XMLName                         xml.Name                        `xml:"AssumeRoleWithWebIdentityResponse"`
+		AssumeRoleWithWebIdentityResult AssumeRoleWithWebIdentityResult `xml:"AssumeRoleWithWebIdentityResult"`
+	}
+
+	respFields := AssumeRoleWithWebIdentityResponse{}
+	err = xml.Unmarshal(bodyBytes, &respFields)
+
+	if respFields.AssumeRoleWithWebIdentityResult.Credentials.SessionToken != "" {
```

Comment:
Can this be refactored to check all variables before setting environment vars. That was the happy path is not indented in.